pkg/cloud/services/ec2/instances.go

```diff
@@ -1138,34 +1145,48 @@ func getCapacityReservationSpecification(capacityReservationID *string) *ec2.Cap
 	}
 }
 
-func getInstanceMarketOptionsRequest(spotMarketOptions *infrav1.SpotMarketOptions) *ec2.InstanceMarketOptionsRequest {
-	if spotMarketOptions == nil {
-		// Instance is not a Spot instance
-		return nil
+func getInstanceMarketOptionsRequest(i *infrav1.Instance) (*ec2.InstanceMarketOptionsRequest, error) {
+	if i.UseCapacityBlock != nil && i.SpotMarketOptions != nil {
+		return nil, errors.New("can't create spot capacity-blocks, remove spot market request")
 	}
 
-	// Set required values for Spot instances
-	spotOptions := &ec2.SpotMarketOptions{}
-
-	// The following two options ensure that:
-	// - If an instance is interrupted, it is terminated rather than hibernating or stopping
-	// - No replacement instance will be created if the instance is interrupted
-	// - If the spot request cannot immediately be fulfilled, it will not be created
-	// This behaviour should satisfy the 1:1 mapping of Machines to Instances as
-	// assumed by the Cluster API.
-	spotOptions.SetInstanceInterruptionBehavior(ec2.InstanceInterruptionBehaviorTerminate)
-	spotOptions.SetSpotInstanceType(ec2.SpotInstanceTypeOneTime)
-
-	maxPrice := spotMarketOptions.MaxPrice
-	if maxPrice != nil && *maxPrice != "" {
-		spotOptions.SetMaxPrice(*maxPrice)
+	// Handle Capacity Block case.
+	if ptr.Deref(i.UseCapacityBlock, false) {
+		if i.CapacityReservationID == nil {
+			return nil, errors.Errorf("capacityReservationID is required when CapacityBlock is enabled")
+		}
+		return &ec2.InstanceMarketOptionsRequest{
+			MarketType: aws.String(ec2.MarketTypeCapacityBlock),
+		}, nil
```

Comment:
Why not make the exposed API mirror this in a way that we can expand the values in the future, without having to have lots of `UseXYZ` kinds of fields?

Would be much better to expose an enum, bools don't age well

Reply:
we have 3 marketType options here: spot, on-demand and capacity-block where we can omit on-demand.

Reply:
@JoelSpeed How about now?
Adding the MarketType API field. It can be set to two values spot/ capacity-block. If not specified and `SpotMarketOptions` is specified default we will set MarketType to `capacity-block`